File: TemplateMethodPattern/src/com/company/transportoPriemones/KelionePrabangiuAutomobiliu.java
package com.company.transportoPriemones;

import com.company.Kelione;
import com.company.kainosSkaiciavimas.Ikainiai;
import com.company.kainosSkaiciavimas.IkainiaiBuilder;

public abstract class KelionePrabangiuAutomobiliu extends Kelione {
    private int priemoneId;

    public KelionePrabangiuAutomobiliu(int priemoneId, String isvykimoTaskas) throws IllegalArgumentException, UnsupportedOperationException {
        super(priemoneId, isvykimoTaskas);

        if (!((priemoneId > 20000) && (priemoneId < 30000)))
            throw new IllegalArgumentException("Nera transporto priemones su tokiu identifikaciniu numeriu.");

        transportoPriemonesPavadinimas = pasirinktiTransportoPriemone(priemoneId);
        if (transportoPriemonesPavadinimas.isEmpty())
            throw new UnsupportedOperationException("Nepavyko pasirinkti transporto priemones.");

        ikainiai = gautiTransportoPriemonesIkainius();
    }

    @Override
    public String pasirinktiTransportoPriemone(int priemoneId) {
        //if(priemoneID nepasirinktas) pasirinkti
        this.priemoneId = priemoneId;
        return "Lexus RC 300";
        //else return null;
    }

    @Override
    public Ikainiai gautiTransportoPriemonesIkainius() {
        return new IkainiaiBuilder()
                .setPradineKaina(300)
                .setKilometroKaina(50)
                .setMinutesKaina(20)
                .setVirsytoLaikoKaina(10)
                .setVirsytoAtstumoKaina(10)
                .setParosKaina(4000)
                .setSavaitesKaina(20000)
                .setMenesioKaina(400000)
                .createIkainiai();
    }

    @Override
    public void paliktiTransportoPriemone() {
        //palikti priemoneId
        System.out.println("Paliktas Lexus RC 300, id: " + priemoneId);
    }

    @Override
    public void pranestiApieNetiketuma() {
        System.out.println("Dekojame uz pranesta Lexus RC 300, id: " + priemoneId + " apgadinima.");
    }
}

File: Mixins/src/com/company/keliones/KelionePrabangiuAutomobiliu.scala
package com.company.keliones

import com.company.kainosSkaiciavimas.Ikainiai

@throws[UnsupportedOperationException]
abstract class KelionePrabangiuAutomobiliu
(override val priemoneId: Int, override val isvykimoTaskas: String) extends Kelione(priemoneId, isvykimoTaskas) {
  transportoPriemonesPavadinimas = pasirinktiTransportoPriemone(priemoneId)
  if (transportoPriemonesPavadinimas.isEmpty) throw new UnsupportedOperationException("Nepavyko pasirinkti transporto priemones.")
  ikainiai = gautiTransportoPriemonesIkainius

  override def pasirinktiTransportoPriemone(priemoneId: Int): String = { //if(priemoneID nepasirinktas) pasirinkti
    "Lexus RC 300"
    //else return null;
  }

  override def gautiTransportoPriemonesIkainius = new Ikainiai(300, 50, 20, 10, 10, 4000, 20000, 400000)

  override def paliktiTransportoPriemone(): Unit = { //palikti priemoneId
    System.out.println("Paliktas Lexus RC 300, id: " + priemoneId)
  }

  override def pranestiApieNetiketuma(): Unit = System.out.println("Dekojame uz pranesta Lexus RC 300, id: " + priemoneId + " apgadinima.")
}